src/ExecutionPathTracer/ExecutionPathTracer.cs

```diff
@@ -97,33 +65,26 @@ public void OnOperationStartHandler(ICallable operation, IApplyData arguments)
         /// </summary>
         public void OnOperationEndHandler(ICallable operation, IApplyData result)
         {
-            this.currentDepth--;
+            if (this.operations.Count <= 1) return;
 
-            // If `compositeTracer` is not null, handle the incoming operation recursively.
-            if (this.compositeTracer != null)
-            {
-                // If the current operation is the composite operation we started with, append
-                // the operations traced out by the `compositeTracer` to the current list of
-                // operations and reset.
-                if (operation == this.currCompositeOp)
-                {
-                    this.AddCompositeOperations();
-                    this.currCompositeOp = null;
-                    this.compositeTracer = null;
-                    return;
-                }
-
-                // Pass operations down to it for handling
-                this.compositeTracer.OnOperationEndHandler(operation, result);
-                return;
-            }
+            var currentOperation = this.operations.Pop();
```

Comment:
This pattern happens enough in your PR, it might help to make it an internal extension method:

```
internal bool TryPop<TSource>(this Stack<TSource> source, out TSource? element)
{
    if (source.Count == 0)
    {
       element = null;
       return false;
    }
    else
    {
        element = source.Pop();
        return true;
    }
}
```
That would allow for replacing this line with something similar to:
```C#
if (!this.operations.TryPop(out var currentOperation) || currentOperation == null) return;
```

Reply:
The `TSource?` gave the error: `A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint.`

Would it also make sense if  I use `Operation` instead of `TSource` specifically since we know that in this namespace we'll only be using a `Stack<Operation>`?

Reply:
Wait, it's an actual method? https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.stack-1.trypeek?view=netcore-3.1

Reply:
Oh, goodness, I should have known that... sorry about the confusion there. Anyway, for that error, `T?` doesn't make sense for an open `T` without a `where` clause to tell it if `T?` means a nullable reference type or shorthand for `Nullable<T>`. This is the danger of writing code without a compiler, sorry.

Reply:
Gotcha, and no worries! Thanks Chris!